Tests/Client/Unit Tests/WebSocket/WebSocketProviderMock.swift

```diff
@@ -0,0 +1,126 @@
+                self.delegate?.websocketDidDisconnect(self, error: nil)
+            }
+            
+            return
+        }
+        
+        DispatchQueue.main.asyncAfter(deadline: .now() + .milliseconds(100)) {
+            self.isConnected = true
+            self.startTimer()
+            self.delegate?.websocketDidConnect(self)
+            
+            // Send the first event with the current user.
+            let event = """
+                {
+                "created_at" : "2020-05-02T13:21:03.862065063Z",
+                "me" : {
+                "id" : "steep-moon-9",
+                "banned" : false,
+                "unread_channels" : 0,
+                "mutes" : [],
+                "last_active" : "2020-05-02T13:21:03.849219Z",
+                "created_at" : "2019-06-05T15:01:52.847807Z",
+                "devices" : [],
+                "invisible" : false,
+                "unread_count" : 0,
+                "channel_mutes" : [],
+                "image" : "https://i.imgur.com/EgEPqWZ.jpg",
+                "updated_at" : "2020-05-02T13:21:03.855468Z",
+                "role" : "user",
+                "total_unread_count" : 0,
+                "online" : true,
+                "name" : "steep-moon-9",
+                "test" : 1
+                },
+                "type" : "health.check",
+                "connection_id" : "
+                """
+            
+            self.connectionId = UUID().uuidString
+            self.sendMessage(event + self.connectionId! + #"", "cid" : "*"}"#)
+        }
+    }
+    
+    func disconnect() {
+        disconnect(error: nil)
+    }
+    
+    func disconnect(error: WebSocketProviderError?) {
+        isConnected = false
+        connectionId = nil
+        stopTimer()
+        delegate?.websocketDidDisconnect(self, error: error)
+    }
+    
+    func sendPing() {
+        guard let connectionId = connectionId else {
+            return
+        }
+        
+        let event = "{\"type\":\"health.check\",\"connection_id\":\""
+            + connectionId
+            + "\",\"cid\":\"*\",\"created_at\":\"2020-05-01T14:45:35.844826426Z\"}"
+        
+        sendMessage(event)
+        startTimer()
+    }
+    
+    func sendMessage(_ message: String) {
+        DispatchQueue.main.asyncAfter(deadline: .now() + .milliseconds(100)) {
+            self.delegate?.websocketDidReceiveMessage(self, message: message)
+        }
+    }
+    
+    func sendStop() {
+        isConnected = false
+        connectionId = nil
+        stopTimer()
+        let error = WebSocketProviderError(reason: "stop", code: 1000, providerType: Self.self, providerError: nil)
+        delegate?.websocketDidDisconnect(self, error: error)
+    }
+    
+    func startTimer() {
+        stopTimer()
+        timer = DispatchSource.makeTimerSource(queue: .main)
+        timer?.schedule(deadline: .now() + .seconds(timeout))
+        timer?.setEventHandler(handler: sendStop)
+        timer?.resume()
```

Comment:
We shouldn't use real timers. Usually in tests, some kind of virtual time is used to make everything run quickly.

Reply:
I know. But WebSocket use an internal timer to keep the connection. So, I think it will be great to add a test for that too. What do you think?

Reply:
You will need to treat the timer as a dependency of `WebSocket` and add the ability to swap it with the virtual-time timer.